---
title: "Troms√∏ Workshop - Pangenome analysis"
author: "Gerry Tonkin-Hill"
date: "`r Sys.Date()`"
output: 
  html_document:
    fig_width: 12
    fig_height: 8
editor_options: 
  chunk_output_type: console
---

```{r setup, include = FALSE}
knitr::opts_chunk$set(fig.width=20, fig.height=12,
                      echo=FALSE, warning=FALSE, message=FALSE,
                      tidy=TRUE)
options(stringsAsFactors = FALSE)
```

## Load libraries

---

```{r}
library(gridExtra)
library(tidyverse)
library(ape)
library(ggtree)
# devtools::install_github("gtonkinhill/panstripe")
library(panstripe)
```

Here, we will analyse the output `panaroo` using the `panstripe` R function. Most of these function can also be used to analyse the output of other pangenome software packages such as Roary and PanX.

Panaroo has already been run for you using the following command.

```
panaroo -i *.gff -o panaroo_out/ -t 40 --clean-mode strict
```

## Load data

First lets load the gene presence/absence matrix using the `read_rtab` function.

---

```{r}
pa <- read_rtab("./data/gene_presence_absence.Rtab")
```

We can also load a phylogeny built using Gubbins.

---

```{r}
tree <- read.tree("./data/core_snps.final_tree.tre")
```

Lets also load some metadata on which isolates are resistant to Penicillin

---

```{r}
resistance_data <- read_tsv("./data/resistance_data.tsv")
```

and finally lets load the GPSC classifications for these genomes.

---

```{r}
gpscs <- read_csv("./data/GPSC_classifications.csv")
```

## Generate some intial plots

Let's start by generating some common pangenome plots. Although I discourage the use of accumulation curves lets try one anyway! This can be done using the  `plot_acc` function.

---

```{r}
panstripe::plot_acc(pa)
```

Now lets plot the presence and absence patterns of each gene that has a decent annotation. We can ignore those genes with either 'group' or 'Gene' in their name. A plot can be generated using the `plot_tree_pa` function.

---

```{r}
keep <- colnames(pa)[(!grepl('.*group.*', colnames(pa))) & (!grepl('.*Gene.*', colnames(pa))) & 
               (colSums(pa)>10) & (colSums(pa)<100)]

panstripe::plot_tree_pa(tree, pa, genes = keep, label_genes = FALSE, cols = 'black', align = FALSE)
```

We can also generate a t-SNE plot similar to the generated by the PANNINI algorithm using the `plot_tsne` function.

---

```{r}
gpscs_vec <-  factor(gpscs$GPSC[match(rownames(pa),  gpscs$Sample)])
plot_tsne(pa, category = gpscs_vec)
```

## Run Panstripe

[Panstripe](https://github.com/gtonkinhill/panstripe) provides some robust alternatives to the common methods for analysing pangenomes. Start by running the `panstripe` function on the complete dataset.

---

```{r}
ps <- panstripe::panstripe(pa, tree)
```

The significance of different drivers of gene gain and loss can be seen by printing the `summary` data frame.

A significant p-value for the `tip` term indicates that there is a different rate of gene exchange at the tips of the phylogeny compared with the internal branches. This is usually driven by annotation errors or highly mobile elements that do not persist long enough to be observed in multiple genomes.

A significant p-value for the `core` term indicates that there is a significant association between the core genome branch length and the number of gene exchange events.

The `depth` term is less interesting but indicates when there is a difference in our ability to detect older gene exchange events.

---

```{r}
ps$summary
```

We can now plot the resulting fit using the `plot_pangenome_fits` function. This is a good alternative the accumulation curve. 

---

```{r}
plot_pangenome_fits(ps)
```

The previous plot considers each branch of the core genome phylogeny separately. We can also look at the cumulative number of gene gain and loss events using the `plot_pangenome_cumulative`. This is similar to the diagnostic plots producded by tempEst when looking at temporal signal in phylogenies.

---

```{r}
plot_pangenome_cumulative(ps)
```

## Comparing pangenomes

Lets compare the gene gain and loss rates between the two largest GPSCs (1 and 66)

---

```{r}
gpsc1 <- gpscs$Sample[gpscs$GPSC==1]
gpsc66 <- gpscs$Sample[gpscs$GPSC==66]

tree_gpsc1 <- drop.tip(tree, tree$tip.label[!tree$tip.label %in% gpsc1])
tree_gpsc66 <- drop.tip(tree, tree$tip.label[!tree$tip.label %in% gpsc66])

pa_gpsc1 <- pa[rownames(pa) %in% gpsc1,]
pa_gpsc66 <- pa[rownames(pa) %in% gpsc66,]
```

Run panstripe seperately for each GPSC.

---

```{r}
ps1 <- panstripe(pa_gpsc1, tree_gpsc1)
ps66 <- panstripe(pa_gpsc66, tree_gpsc66)
```

Plot the resulting fits on the same plot using `plot_pangenome_fits`

---

```{r}
plot_pangenome_fits(list(gpsc1=ps1, gpsc66=ps66))
```

We can also compare the resulting parameter estimates visually using the `plot_pangenome_params` function.

---

```{r}
plot_pangenome_params(list(gpsc1=ps1, gpsc66=ps66))
```

To more formally test whether the parameter estimate differ run the `compare_pangenomes` function.

---

```{r}
cps <- compare_pangenomes(ps1, ps66)
cps$summary
```

## Lineage specific genes

Recently, researchers have started to look for genes that are specific for a particular lineage. Start by identifying those genes that are present in all GPSC1 isolates and not at all in GPSC66.

---

```{r}
perc1 <- colSums(pa_gpsc1)/nrow(pa_gpsc1)

perc66 <- colSums(pa_gpsc66)/nrow(pa_gpsc66)

colnames(pa_gpsc1)[perc1>=1 & perc66<=0]
```

Although this is becoming a common thing to do I am unsure how we can interpret results such as these. If everyone gets to this point we can have a group discussion about how to define a lineage specific gene and when this is interesting.
