---
title: "Troms√∏ Workshop - Transmission"
author: "Gerry Tonkin-Hill"
date: "`r Sys.Date()`"
output: 
  html_document:
    fig_width: 12
    fig_height: 8
editor_options: 
  chunk_output_type: console
---

```{r setup, include = FALSE}
knitr::opts_chunk$set(fig.width=20, fig.height=12,
                      echo=FALSE, warning=FALSE, message=FALSE,
                      tidy=TRUE)
options(stringsAsFactors = FALSE)
```

## Load libraries

```{r}
library(gridExtra)
library(tidyverse)
library(ape)
library(ggtree)
# devtools::install_github("xavierdidelot/BactDating")
library(BactDating)
# devtools::install_github('xavierdidelot/TransPhylo')
library(TransPhylo)
# devtools::install_github("ChrispinChaguza/RCandy")
library(RCandy)
# devtools::install_github("JamesStimson/transcluster")
library(transcluster)
```


## Load the data

We can load the core SNP alignment using the ape package.

---

```{r}
core_aln <- read.dna("./data/core_snps.fasta", format = 'fasta', as.matrix = TRUE, as.character = TRUE)
```

I have also included a csv file with the Global Pneumococcal Sequencing Cluster (GPSC) classifications to help with visualisation. This can be loaded using the following command.

---

```{r}
gpscs <- read_csv("./data/GPSC_classifications.csv")
```

We can also load a phylogeny built using Gubbins.

---

```{r}
tree <- read.tree("./data/core_snps.final_tree.tre")
```

Lets also load some metadata on which isolates are resistant to Penicillin

---

```{r}
resistance_data <- read_tsv("./data/resistance_data.tsv")
```

and finally the dates at which the genomes were sampled

---

```{r}
dates <- read_csv("./data/dates.csv")
```

## Visualise recombination with RCandy

Use the [RCandyVis](https://github.com/ChrispinChaguza/RCandy) function to visualise the recombination tracts inferred using Gubbins. You will need the recombination predictions `./data/core_snps.recombination_predictions.gff` and the annotated reference genome `./data/Streptococcus_pneumoniae_ATCC_700669.gff`.

---

```{r}
RCandyVis(tree.file.name = "./data/core_snps.final_tree.tre", 
          midpoint.root = TRUE, ladderize.tree.right = TRUE, 
          taxon.metadata.file = "./data/resistance_data.csv", taxon.metadata.columns = c("Resistance"), 
          taxon.id.column = "Sample", 
          gubbins.gff.file = "./data/core_snps.recombination_predictions.gff",
          ref.genome.name = "./data/Streptococcus_pneumoniae_ATCC_700669.gff")
```

## Transmission

We will now investigate transmission between the samples.

### Pairwise SNP distance

Starting with the most simple approach, first calculate the SNP distance between genomes using ape's `read.dna` and `dist.dna` functions.

---

```{r}
fasta_aln <- read.dna("./data/core_snps.fasta", format = 'fasta')
d <- dist.dna(fasta_aln, model = "N")
```

Plot the distribution of pairwise distances using the `hist` function

---

```{r}
hist(c(d), xlab = 'SNP distance', main='Histogram')
```

A common SNP distance threshold for identifying potential transmission links is 12 in Streptococcus pneumoniae. Cluster the isolates if they are within 12 SNPs using single linkage clustering. This can be achieved using the `hclust` and `cutree` functions.

---

```{r}
h <- hclust(d, method = 'single')
clusters <- cutree(h, h = 12)
clusters <- tibble(Sample=names(clusters), cluster=clusters)
```

Plot the results against the phylogeny using `ggtree` and `facet_plot`

---

```{r}
gg <- ggtree(tree)

f2 <- facet_plot(gg, panel = "clusters", data = clusters, geom = geom_tile, aes(x = cluster), 
    color = "blue")
f2
```

## Transcluster

So far we have completely ignored the date samples were taken. We can improve this using the TransCluster function. Here we will use the original R package but if you are clustering very large alignments [fasttranscluster](https://github.com/gtonkinhill/fasttranscluster) can be a good alternative.

You will need to convert the dates to numeric years prior to running TransCluster.

---

```{r}
years_from_earliest <- as.numeric((dates$Date - lubridate::as_date("2007-12-30"))/365)
snpMatrix <- as.matrix(d)
ids <- rownames(snpMatrix)
```

Run [TransCluster](https://github.com/JamesStimson/transcluster). We assume a generation time of 2 months (6 per/year) and a substitution rate of 5.3 snps/year. You will need the `createModel`, `setParams` and `setTransThresholds` functions to set up the model.

---

```{r}
myModel <- createModel(ids, years_from_earliest, snpMatrix)
myModel <- setParams(myModel, lambda=5.3, beta=6)
myModel <- setTransThresholds(myModel, c(1,5,10))
```

Generate the clusters using the `setCutoffs` and `makeTransClusters` functions.

---

```{r}
myModel <- setCutoffs(myModel)
transclusters <- makeTransClusters(myModel)
```

Plot the results against the phylogeny using `ggtree` and `facet_plot`. 

---

```{r}
gg <- ggtree(tree)

tclustlevel <- tibble(Sample = myModel$id, 
                      `Within 1`=transclusters$`1`, 
                      `Within 5`=transclusters$`5`,
                      `Within 10`=transclusters$`10`)

f1 <- facet_plot(gg, panel = "Within 1", data = tclustlevel, geom = geom_tile, aes(x = `Within 1`), 
    color = "blue")
f2 <- facet_plot(f1, panel = "Within 5", data = tclustlevel, geom = geom_tile, aes(x = `Within 5`), 
    color = "red")
f3 <- facet_plot(f2, panel = "Within 10", data = tclustlevel, geom = geom_tile, aes(x = `Within 10`), 
    color = "green")
f3
```

## Bactdater

So far we have only considered pairwise relationships. We can often do better than this by building a phylogeny. To account for the time saples were taken we need to create a dated tree. 

Here we focus on a smaller cluster where we are more likely to get an accurate dating of the tree.

---

```{r}
subtree <- ape::extract.clade(tree, 321)
```

---

```{r}
sorted_dates <- dates$Date[match(subtree$tip.label, dates$Sample)]
max_date <- max(sorted_dates)
sorted_dates <- as.numeric((sorted_dates - lubridate::as_date("2007-12-30"))/365)
```

Use the `roottotip` function to test for temporal signal

---

```{r}
r=roottotip(subtree, as.numeric(sorted_dates))
```

Now run Nactdater using the `bactdate` function and plot the result.

---

```{r}
res=bactdate(subtree, sorted_dates)
plot(res,'treeCI')
```

So far we have ignored the fact that Gubbins has filtered out parts of the alignment to deal with recombination. Use the `loadGubbins` function to re-run the analysis taking recombination into account. Use the `drop.tip.useRec` function to filter the tree to the same set of samples. You will need to use the `./data/core_snps` prefix to do this.

---

```{r}
gubbins <- loadGubbins("./data/core_snps")

gubbins <- drop.tip.useRec(gubbins, tree$tip.label[!tree$tip.label %in% subtree$tip.label])

res2=bactdate(gubbins, sorted_dates, useRec=TRUE)

plot(res2,'treeCI')
```

## TransPhylo

Now we have a dated tree we can infer a transmission network using [TransPhylo](https://github.com/xavierdidelot/TransPhylo). As with TransCluster we will assume a mean generation time of 2 months (2/12 years) and a standard deviation of 1.

First convert the phylo object into the correct format using the `ptreeFromPhylo` function and generate a plot. We will assume that the last sample was taken in 2010.

---

```{r}
ptree <- ptreeFromPhylo(res2$tree, dateLastSample=2010)
plot(ptree)
```

We can now infer the transmission network using the `inferTTree` function. It is often better to use the slower starting function (optiStart=1) and a good starting point for the number of mcmcInterations is 10,000. Let's assume the date sampling finished was 2010.5.

---

```{r}
tres <- inferTTree(ptree, mcmcIterations=1000, w.mean=2/12, w.std=1, dateT=2010.5, optiStart = 1)
```

Investigate whether the algorithm has converged by plotting the resulting object. We are looking to see that the MCMC chains have converged to consistent values.

---

```{r}
plot(tres)
```

Now plot the resulting transmission networks using the `medTTree` and `extractTTree` functions.

---

```{r}
med=medTTree(tres)
plot(med)

ttree=extractTTree(med)
plot(ttree, w.shape = 2/12, w.scale = 1)
```

Finally, generate a pairwise transmission matrix from the inferred network using the `computeMatTDist` and `levelplot` functions.

---

```{r}
mat=computeMatTDist(tres)
lattice::levelplot(mat,xlab='',ylab='')
```

